### ISSUE TITLE
[Enhacement] [x:Bind] Strongly bind to ViewModel instance

### ISSUE THREAD

**Issue by andreinitescu** [xamarin/Xamarin.Forms#3030]
### Rationale 

Support for x:Bind should bring the possibility to **bind without reflection**:

MainPage.xaml:

`<Label Text="{x:Bind ViewModel.Username}" />`

MainPage.xaml.cs:
```
public partial class MainPage : ContentPage
{
    public MainPageViewModel ViewModel { get; }

    public MainPage()
    {
        InitializeComponent();

        BindingContextChanged += (s, e) =>
        {
            ViewModel = BindingContext as MainPageViewModel;
        };
    }
}
```

XAMLC will emit at compile time code similar to the following:

```
var l1 = new Label();
var b1 = new StronglyTypedBinding("Username", () => ViewModel.UserName);
l1.SetBinding(Label.TextProperty, b1);

```

Where `StronglyTypedBinding` is a new `Binding` derived class which knows to listen to `PropertyChanged` on `((INotifyPropertyChanged)ViewModel)` for "Username" property name and call `ViewModel.UserName` to read the value (or even write to to it if binding is two-way)

Binding without reflection should bring some performance improvements when rendering pages, the performance gain obviously depends on the complexity of the page layout.

**Comment by StephaneDelcroix** [comment]
`x:Bind` particular syntax is coming, but XamlC already allows you to compile Bindings. specify the x:DataType, and you're done.

It generates code similar to yours for OneTime bindings, but also generates code for 1Way 2Ways, and even can hook into INotifyPropertyChanged

given the following xaml

```xaml
<StackLayout x:DataType="local:MockViewModel">
    <Label Text="{Binding Text, Mode=OneTime}" x:Name="label8" />
</StackLayout>
```

this is the code generated

```csharp
label8.SetBinding(Label.TextProperty, new TypedBinding<MockViewModel, string>((MockViewModel mockViewModel) => mockViewModel.Text, null, null);
```

**Comment by adrianknight89** [comment]
@StephaneDelcroix so, if we use `x:Bind` in the future, then we don't have to specify `x:DataType`?

**Comment by andreinitescu** [comment]
@StephaneDelcroix So you're saying binding **without any reflection involved** is possible today by using `x:DataType` ? 

To be clear, I am interested about binding without using reflection **at all**.

I made a sample app with this XAML:

```
<ContentPage xmlns="http://xamarin.com/schemas/2014/forms"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:local="clr-namespace:App20"
             x:Class="App20.MainPage"
             BackgroundColor="BlanchedAlmond"
             x:DataType="local:MainPageViewModel">
    <StackLayout VerticalOptions="Center"
                 HorizontalOptions="Center">
        <Label Text="{Binding Item.Name}" />
        <Label Text="{Binding Item.ChildItem.Name}" />
    </StackLayout>
</ContentPage>

```

The emitted code seems to be using `BindingExtension` and it's using the string binding path "Item.Name":

![snip_20180615104027](https://user-images.githubusercontent.com/743918/41456347-d857f740-7088-11e8-80bf-beb8b2a1506d.png)

But looking at https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Xaml/MarkupExtensions/BindingExtension.cs#L23 I assume the Path string is not actually being used. Correct?

**Comment by StephaneDelcroix** [comment]
the Path is not actually used, the only part being used is the TypedBinding part, which doesn't use the `Item.Name` string, but a delegate instead.

no reflection AT ALL at runtime.

**Comment by StephaneDelcroix** [comment]
@adrianknight89 x:Bind binds to the root object 🤦‍♂️, and that object type is known. This is the reason why it doesn't require x:DataType (unless you're in a DataTemplate scenario, and then x:Bind binds to the view model 🤷‍♂️ 🤦‍♂️ x2).

the only somehow valid usage for x:Bind is binding to page elements (instead of doing x:Reference), because any other use case goes against proper mvvm design.

**Comment by andreinitescu** [comment]
@StephaneDelcroix Instead of using Tuple which allocates on heap, isn't KeyValuePair a better choice

**Comment by AtosNicoS** [comment]
I tried using `DataType` as suggested, but I have the same problem as described here:
https://forums.xamarin.com/discussion/106017/x-datatype-doesnt-exist

The entry is marked red, the intellisense does not work properly and does not notify me on invalid bindings. Only at compile time I get an error at least but thats still not an ideal solution.

I only found this project from the roadmap:
https://github.com/xamarin/Xamarin.Forms/projects/7

Is there any progress on this implementation?
